### PR TITLE
nn.EmbeddingBag to compute a bag of word embeddings (Embedding + Sum)

### PR DESCRIPTION
**Pull request by soumith**
Computes sums or means of 'bags' of embeddings, without instantiating the intermediate embeddings.

For bags of constant length,

* nn.EmbeddingBag with `mode=sum` is equivalent to nn.Embedding followed by `torch.sum(dim=1)`
* with `mode=mean` is equivalent to nn.Embedding followed by `torch.mean(dim=1)`

However, nn.EmbeddingBag is much more time and memory efficient than using a chain of these operations.

```python
    >>> # an Embedding module containing 10 tensors of size 3
    >>> embedding_sum = nn.EmbeddingBag(10, 3, mode='sum')
    >>> # a batch of 2 samples of 4 indices each
    >>> input = Variable(torch.LongTensor([1,2,4,5,4,3,2,9]))
    >>> offsets = Variable(torch.LongTensor([0,4]))
    >>> embedding_sum(input, offsets)

    Variable containing:
    -0.7296 -4.6926  0.3295
    -0.5186 -0.5631 -0.2792
    [torch.FloatTensor of size 2x3]
```

This diff introduces an efficient GPU implementation, and a slow/reference CPU implementation.
In a subsequent diff, I'll introduce the fast CPU implementation as well.

The initial part of this diff was written by @adamlerer . I changed the naming, introduced mean reduction and added additional tests / docs.

### CODE REVIEW

**Comment by soumith** [comment]
i'm looking to get this in today to unblock @arthurszlam

File: torch/nn/_functions/thnn/sparse.py [hunk 1]
```diff
@@ -103,3 +103,152 @@ def backward(self, grad_output):
...
+        super(EmbeddingBag, self).__init__()
+        self.max_norm = max_norm
+        self.norm_type = norm_type
+        self.scale_grad_by_freq = scale_grad_by_freq
+        self._indices = None
+        assert mode is not None
+        if mode == 'sum':
+            self.mode = 0
+        elif mode == 'mean':
+            self.mode = 1
```
Comment (apaszke) [inline torch/nn/_functions/thnn/sparse.py]: can we please not do this? I can't see any value in converting readable mode names to numbers that don't have any meaning. If you want to do this conversion only once, define them as constants:
```python
MODE_SUM = 0
MODE_MEAN = 1
```

File: torch/nn/_functions/thnn/sparse.py [hunk 2]
```diff
@@ -103,3 +103,152 @@ def backward(self, grad_output):
...
+        self._backend.LookupTable_renorm(
+            self._backend.library_state,
+            indices,
+            weight,
+            self.max_norm,
+            self.norm_type
+        )
+
+    def forward(self, indices, offsets, weight):
+        assert indices.dim() <= 2
```
Comment (apaszke) [inline torch/nn/_functions/thnn/sparse.py]: Don't we only support batched inputs?
Also, use `if`s and raise proper errors instead of asserts

File: torch/nn/_functions/thnn/sparse.py [hunk 3]
```diff
@@ -103,3 +103,152 @@ def backward(self, grad_output):
...
+        self.norm_type = norm_type
+        self.scale_grad_by_freq = scale_grad_by_freq
+        self._indices = None
+        assert mode is not None
+        if mode == 'sum':
+            self.mode = 0
+        elif mode == 'mean':
+            self.mode = 1
+        else:
+            raise InputError("mode needs to be 'sum' or 'mean', but got {}" % mode)
```
Comment (apaszke) [inline torch/nn/_functions/thnn/sparse.py]: This should be `ValueError`

File: torch/nn/_functions/thnn/sparse.py [hunk 4]
```diff
@@ -103,3 +103,152 @@ def backward(self, grad_output):
...
+
+        self._backend = type2backend[type(weight)]
+        self._weight_size = weight.size()
+        self._offset2bag = offsets.new()
+
+        if not indices.is_contiguous():
+            self._indices = indices.contiguous()
+            indices = self._indices
+        else:
+            self.save_for_backward(indices)
```
Comment (apaszke) [inline torch/nn/_functions/thnn/sparse.py]: Can't we just always save them, and do `contiguous()` both here and in backward? That's what all other layers do

File: torch/nn/_functions/thnn/sparse.py [hunk 5]
```diff
@@ -103,3 +103,152 @@ def backward(self, grad_output):
...
+
+        output = weight.new()
+        if self.max_norm is not None:
+            self._renorm(indices, weight)
+
+        assert indices.dim() == 1
+        assert offsets.dim() == 1
+        assert offsets[0] == 0
+        assert offsets[-1] < indices.size(0)
```
Comment (apaszke) [inline torch/nn/_functions/thnn/sparse.py]: raise readable errors

File: torch/nn/_functions/thnn/sparse.py [hunk 6]
```diff
@@ -103,3 +103,152 @@ def backward(self, grad_output):
...
+            self._offset2bag.index_fill_(0, offsets, 1)  # offset2bag = [1 0 1 0 1]
+            self._offset2bag[0] = 0  # offset2bag = [0 0 1 0 1]
+            self._offset2bag = self._offset2bag.cumsum(0)  # offset2bag = [0 0 1 1 2]
+            output.resize_(offsets.size(0), weight.size(1)).zero_()
+            output.index_add_(0, self._offset2bag, index_output)
+            if self.mode == 1:
+                self.sequence_length = weight.new().resize_(offsets.size())
+                self.sequence_length[:-1] = offsets[1:] - offsets[:-1]
+                self.sequence_length[-1] = indices.size(0) - offsets[-1]
+                self.sequence_length = self.sequence_length[:, None].expand_as(output)
```
Comment (apaszke) [inline torch/nn/_functions/thnn/sparse.py]: is `sequence_length` the same as a vector of bag sizes? maybe it's a better name for it

File: torch/nn/_functions/thnn/sparse.py [hunk 7]
```diff
@@ -103,3 +103,152 @@ def backward(self, grad_output):
...
+
+    def backward(self, grad_output):
+        if self._indices is not None:
+            indices = self._indices
+        else:
+            indices, = self.saved_tensors
+
+        grad_output = grad_output.contiguous()
+        if indices.dim() == 2:
+            indices = indices.view(-1)
```
Comment (apaszke) [inline torch/nn/_functions/thnn/sparse.py]: I don't think it's necessary. We should require them to be 2D

**Comment by soumith** [comment]
I've addressed all the review comments.

File: torch/nn/_functions/thnn/sparse.py [hunk 8]
```diff
@@ -103,3 +103,155 @@ def backward(self, grad_output):
...
+        )
+
+    def forward(self, weight, indices, offsets):
+        assert not self.needs_input_grad[1], "EmbeddingBag doesn't " \
+            "compute the gradient w.r.t. the indices"
+
+        assert not self.needs_input_grad[2], "EmbeddingBag doesn't " \
+            "compute the gradient w.r.t. the offsets"
+
+        assert indices.dim() == 1
```
Comment (apaszke) [inline torch/nn/_functions/thnn/sparse.py]: error message?

File: torch/nn/_functions/thnn/sparse.py [hunk 9]
```diff
@@ -103,3 +103,155 @@ def backward(self, grad_output):
...
+        indices = indices.contiguous().view(-1)
+        output = weight.new()
+        if self.max_norm is not None:
+            self._renorm(indices, weight)
+
+        if weight.is_cuda:
+            if self.mode == MODE_MEAN:
+                self.sequence_length = offsets.new().resize_(offsets.size())
+            else:
+                self.sequence_length = None
```
Comment (apaszke) [inline torch/nn/_functions/thnn/sparse.py]: can we really do s/sequence_length/bag_sizes/? nothing else in this module, nor in its docs mentions sequences

**Review by apaszke — APPROVED** [review]
Looks good. You can optionally fix the two things I found now. I didn't look into the C code

File: torch/nn/_functions/thnn/sparse.py (hunk 8 above)
Comment (soumith) [inline torch/nn/_functions/thnn/sparse.py]: doesn't need it here, as the Module code does the error message.

File: torch/nn/_functions/thnn/sparse.py (hunk 9 above)
Comment (soumith) [inline torch/nn/_functions/thnn/sparse.py]: okay changing. I also found via contbuild that the corner-case of 1 sequence (i.e. offsets.size(0) == 1) is not handled on the CPU side. I'll fix that as well.

**Comment by soumith** [comment]
this is now merged into master
